src/core/cluster.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Collection of global cluster state for the ZooKeeper quorum."""
+import logging
+from typing import Dict, Optional, Set
+
+from charms.data_platform_libs.v0.data_interfaces import DataPeerOtherUnit
+from ops.framework import Framework, Object
+from ops.model import Relation, Unit
+
+from core.models import (
+    SUBSTRATES,
+    CharmWithRelationData,
+    ODCluster,
+    ODServer,
+    OpensearchServer,
+)
+from literals import OPENSEARCH_REL_NAME, PEER
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterStateBase(Object):
+    """Collection of global cluster state for Framework/Object."""
+
+    def __init__(self, charm: Framework | Object, substrate: SUBSTRATES):
```

Comment:
I would prefer not to restrain that for now the least.

Since we take no advantage of passing the charm (as a heir for `CharmBase`), but all we need is the `model`, I would prefer to reflect that in the type specification (and perhaps even alter the cdde later on, not to pass the charm...?)